File: iterations/tabletest/iteration-45/eval-28-convert-from-methodsource/outputs/src/test/java/com/example/ShippingCostCalculatorTest.java
package com.example;

import org.junit.jupiter.api.DisplayName;
import org.tabletest.junit.Description;
import org.tabletest.junit.TableTest;
import org.tabletest.junit.TypeConverter;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class ShippingCostCalculatorTest {

    private final ShippingCostCalculator calculator = new ShippingCostCalculator();

    @DisplayName("Determines base cost from shipping zone, weight, and dimensional weight")
    @Description("""
        Package options are absent (no fragile, insured, or hazmat handling) in every row.
        Carrier is fixed at DHL; carrier choice does not affect cost
        (see costIsUnaffectedByCarrier).
        """)
    @TableTest("""
        Scenario                                  | Zone        | Weight (kg) | Dimensions (cm) | Cost?
        Light package                             | EU standard | 0.5         | [20, 15, 10]    | 5.00
        Medium package                            | EU standard | 3.0         | [30, 20, 15]    | 7.50
        Heavy package                             | EU standard | 10.0        | [40, 30, 20]    | 12.50
        Very heavy package                        | EU standard | 25.0        | [50, 40, 30]    | 20.00
        Light package, express                    | EU express  | 0.5         | [20, 15, 10]    | 8.00
        Medium package, express                   | EU express  | 3.0         | [30, 20, 15]    | 12.00
        Light package, US standard                | US standard | 0.5         | [20, 15, 10]    | 7.00
        Heavy package, US express                 | US express  | 10.0        | [40, 30, 20]    | 30.00
        Dimensional weight exceeds actual weight  | EU standard | 1.0         | [70, 50, 10]    | 12.50
        """)
    void resolvesBaseCost(ShippingZone zone, double weight, List<Integer> dimensions, BigDecimal cost) {
        BigDecimal result = calculator.calculateShippingCost(zone, weight, dimensions, new PackageOptions(), Carrier.DHL);
        assertEquals(0, result.compareTo(cost));
    }

    @DisplayName("Applies fragile, insurance, hazmat, and oversize surcharges to the base cost")
    @Description("""
        All rows use an EU standard zone at 3.0 kg actual weight, where the base cost
        before surcharges is 7.50 (see resolvesBaseCost). Carrier is fixed at DHL and
        does not affect cost (see costIsUnaffectedByCarrier). The fragile-and-insured row
        shows that the fragile multiplier applies to the base cost before the insurance
        premium is added.
        """)
    @TableTest("""
        Scenario             | Dimensions (cm) | Options                            | Cost?
        No surcharges        | [30, 20, 15]    | [:]                                | 7.50
        Fragile surcharge    | [30, 20, 15]    | [fragile: true]                    | 8.625
        Insured surcharge    | [30, 20, 15]    | [insuredValue: 500]                | 10.50
        Fragile and insured  | [30, 20, 15]    | [fragile: true, insuredValue: 200] | 11.625
        Hazmat handling      | [30, 20, 15]    | [handling: hazmat]                 | 15.50
        Oversize dimension   | [120, 5, 5]     | [:]                                | 17.50
        """)
    void appliesSurchargesToBaseCost(List<Integer> dimensions, PackageOptions options, BigDecimal cost) {
        ShippingZone zone = new ShippingZone("EU", "standard");
        BigDecimal result = calculator.calculateShippingCost(zone, 3.0, dimensions, options, Carrier.DHL);
        assertEquals(0, result.compareTo(cost));
    }

    @DisplayName("Carrier choice does not affect shipping cost")
    @TableTest("""
        Scenario                     | Zone        | Weight (kg) | Dimensions (cm) | Carrier           | Cost?
        Carrier does not affect cost | EU express   | 3.0        | [30, 20, 15]    | {DHL, UPS, FEDEX} | 12.00
        """)
    void costIsUnaffectedByCarrier(ShippingZone zone, double weight, List<Integer> dimensions, Carrier carrier, BigDecimal cost) {
        BigDecimal result = calculator.calculateShippingCost(zone, weight, dimensions, new PackageOptions(), carrier);
        assertEquals(0, result.compareTo(cost));
    }

    @TypeConverter
    public static ShippingZone parseZone(String value) {
        String[] parts = value.split(" ");
        return new ShippingZone(parts[0], parts[1]);
    }

    @TypeConverter
    public static PackageOptions parseOptions(Map<String, String> config) {
        PackageOptions options = new PackageOptions();
        if ("true".equals(config.get("fragile"))) {
            options.setFragile(true);
        }
        if (config.containsKey("insuredValue")) {
            options.setInsuredValue(new BigDecimal(config.get("insuredValue")));
        }
        if (config.containsKey("handling")) {
            options.setHandling(config.get("handling"));
        }
        return options;
    }
}

File: iterations/tabletest/next/iteration-8/eval-26-convert-from-kotest/outputs/src/test/kotlin/com/example/ShippingCostCalculatorTest.kt
package com.example

import org.junit.jupiter.api.Assertions.assertEquals
import org.tabletest.junit.Description
import org.tabletest.junit.TableTest
import org.tabletest.junit.TypeConverter
import java.math.BigDecimal

@TypeConverter
fun parsePackageOptions(config: Map<String, String>): PackageOptions {
    val options = PackageOptions()
    config["fragile"]?.let { options.isFragile = it.toBoolean() }
    config["insuredValue"]?.let { options.insuredValue = BigDecimal(it) }
    config["handling"]?.let { options.handling = it }
    return options
}

class ShippingCostCalculatorTest {

    private val calculator = ShippingCostCalculator()

    @Description("""
        Dimensions are fixed to [10, 10, 10] in every row -- well below the dimensional-weight
        threshold, so effective weight always equals actual weight here. Carrier is fixed to
        DHL; carrier choice does not affect cost (see carrierDoesNotAffectCost).
        """)
    @TableTest("""
        Scenario                     | Region | Speed    | Weight | Base Cost?
        EU standard, light tier      | EU     | standard | 0.5    | 5.00
        EU standard, medium tier     | EU     | standard | 3.0    | 7.50
        EU standard, heavy tier      | EU     | standard | 10.0   | 12.50
        EU standard, very heavy tier | EU     | standard | 25.0   | 20.00
        EU express, light tier       | EU     | express  | 0.5    | 8.00
        EU express, medium tier      | EU     | express  | 3.0    | 12.00
        US standard, light tier      | US     | standard | 0.5    | 7.00
        US express, heavy tier       | US     | express  | 10.0   | 30.00
        """)
    fun determinesBaseRateByRegionSpeedAndWeight(region: String, speed: String, weight: Double, baseCost: BigDecimal) {
        val zone = ShippingZone(region, speed)
        val cost = calculator.calculateShippingCost(zone, weight, listOf(10, 10, 10), PackageOptions(), Carrier.DHL)
        assertEquals(0, baseCost.compareTo(cost))
    }

    @Description("""
        Zone is fixed to EU standard. Effective weight is the greater of actual weight and
        dimensional weight (volume / 5000, rounded to 3 decimal places).
        """)
    @TableTest("""
        Scenario                     | Weight | Dimensions   | Base Cost?
        Actual weight is larger      | 25.0   | [50, 40, 30] | 20.00
        Dimensional weight is larger | 1.0    | [70, 50, 10] | 12.50
        """)
    fun selectsEffectiveWeightFromActualOrDimensional(weight: Double, dimensions: List<Int>, baseCost: BigDecimal) {
        val zone = ShippingZone("EU", "standard")
        val cost = calculator.calculateShippingCost(zone, weight, dimensions, PackageOptions(), Carrier.DHL)
        assertEquals(0, baseCost.compareTo(cost))
    }

    @Description("""
        Zone is fixed to EU standard, weight to 3.0kg (base rate 7.50 with dimensions
        [30, 20, 15]). Surcharges apply in order: oversize fee, then hazmat fee, then the
        fragile multiplier, then the insurance premium -- added last, so it is unaffected by
        the fragile multiplier.
        """)
    @TableTest("""
        Scenario             | Dimensions   | Options                            | Total Cost?
        No surcharges        | [30, 20, 15] | [:]                                | 7.50
        Oversize dimension   | [120, 5, 5]  | [:]                                | 17.50
        Fragile              | [30, 20, 15] | [fragile: true]                    | 8.625
        Insured              | [30, 20, 15] | [insuredValue: 500]                | 10.50
        Fragile and insured  | [30, 20, 15] | [fragile: true, insuredValue: 200] | 11.625
        Hazmat handling      | [30, 20, 15] | [handling: hazmat]                 | 15.50
        """)
    fun appliesSurchargesToBaseCost(dimensions: List<Int>, options: PackageOptions, totalCost: BigDecimal) {
        val zone = ShippingZone("EU", "standard")
        val cost = calculator.calculateShippingCost(zone, 3.0, dimensions, options, Carrier.DHL)
        assertEquals(0, totalCost.compareTo(cost))
    }

    @Description("""
        Fixed for all rows: zone = EU express, weight = 3.0kg, dimensions = [30, 20, 15]
        (base rate 12.00, no other surcharges).
        """)
    @TableTest("""
        Scenario                            | Carrier           | Total Cost?
        Carrier choice does not affect cost | {DHL, UPS, FEDEX} | 12.00
        """)
    fun carrierDoesNotAffectCost(carrier: Carrier, totalCost: BigDecimal) {
        val zone = ShippingZone("EU", "express")
        val cost = calculator.calculateShippingCost(zone, 3.0, listOf(30, 20, 15), PackageOptions(), carrier)
        assertEquals(0, totalCost.compareTo(cost))
    }
}
